### PR TITLE
Copies notes details from first comments to notes table

### PR DESCRIPTION
**Pull request by nenad-vujicic**
### Description

PR adds migration script for copying notes descriptions, user_ids and user_ips from notes first comments to notes table records. In 2 commits we propose 2 variants of migration script (not sure which is more appropriate for the situation because of large number of notes):

1. basic version - for every `batch of notes`, dynamically generates SQL query for copying values and executes it (simpler to understand, ..).
2. enhanced version - same as previous but keeps track about notes for which migration failed (for some reason?) and retries again later (more robust, ..)

### How has this been tested?

Automated and manual tests before and after data migrations (both basic and enhanced one)

### CODE REVIEW

**Comment by tomhughes** [comment]
I think this will do the job, obviously with batches of note IDs in the CTE expression:

```sql
WITH first_comment AS(
  SELECT DISTINCT ON (note_id) *
  FROM note_comments
  WHERE note_id BETWEEN 1000 AND 2000
  ORDER BY note_id, id
)
UPDATE notes
SET description = first_comment.body,
    user_id = first_comment.author_id,
    user_ip = first_comment.author_ip
FROM first_comment
WHERE first_comment.note_id = notes.id
  AND first_comment.event = 'opened';
```

**Comment by nenad-vujicic** [comment]
> I think this will do the job, obviously with batches of note IDs in the CTE expression:
> 

Thanks! I think I fixed it (with the latest push). Here is the code:
```ruby
Note.in_batches(:of => 1000) do |notes|
  note_ids = notes.pluck(:id)

  sql_query = <<-SQL.squish
    WITH first_comment AS(
      SELECT DISTINCT ON (note_id) *
      FROM note_comments
      WHERE note_id BETWEEN #{note_ids.min} AND #{note_ids.max}
      ORDER BY note_id, id
    )
    UPDATE notes
    SET description = first_comment.body,
        user_id = first_comment.author_id,
        user_ip = first_comment.author_ip
    FROM first_comment
    WHERE first_comment.note_id = notes.id
      AND first_comment.event = 'opened';
  SQL

  ActiveRecord::Base.connection.execute(sql_query)
end
```

**Comment by gravitystorm** [comment]
In my own projects, we use a different approach for data migrations.

* First, we avoid using rails built-in database migrations. These are great for structure changes, but not for data changes, since they can't be tested. And there's often some edge cases that we want to have confidence about before we run the data migration and tests are the foundation for confidence.
* We create a job to migrate the data, so that it runs in a background process. We have a rake task that simply queues the job and nothing else. This is because rake tasks are convenient for us to run, but again are very hard to test. Jobs are easy to test.
* We have a separate queue for slow jobs, with a separate set of workers, so that if it takes e.g. 10min to run we are still sending out confirmation emails in the meantime.
* For the task itself, we write it in a way that it can be retried multiple times. So for example we only select the objects that have not yet been migrated (e.g. `Note.where(description: nil).in_batches`) so that if the job is retried, it starts from where it left off, not from the start again.

We do this because in so many cases either we forgot about an edge case but picked it up in testing, or there turns out to be some unexpected data that messes things up (an example would be a note with no comments, even if there's no way that could have happened, in theory, somehow it always does). It also avoid any worries about locks, long-running transactions, transactions with large amounts of uncommitted changes, timeouts, network interruptions, db process restarts and so on.

I say all this to provide an alternative approach that I know works. But in this case, since @tomhughes will be running the actual migration process, whatever approach he prefers is what we should do!

File: db/migrate/20250212160355_backfill_note_descriptions.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+          WHERE note_id BETWEEN #{note_ids.min} AND #{note_ids.max}
+          ORDER BY note_id, id
+        )
+        UPDATE notes
+        SET description = first_comment.body,
+            user_id = first_comment.author_id,
+            user_ip = first_comment.author_ip
+        FROM first_comment
+        WHERE first_comment.note_id = notes.id
+          AND first_comment.event = 'opened';
```
Comment (gravitystorm) [inline db/migrate/20250212160355_backfill_note_descriptions.rb:24]: Why is this restricted to only comments with event = 'opened'? That's not the same logic as is currently used for creating note descriptions using comments - it uses `all_comments.first` without any similar restriction.

**Review by gravitystorm — COMMENTED** [review]
I would also interested in what the next step will be. How do you propose removing the first comments, when they are no longer used? That migration will need to work in an "online" manner, i.e. there needs to be code that runs before the migration, during the migration and after the migration. How will that code know if the first comment has been already deleted for a given note? It can't just be `all_comments.drop(1)` for example, since that won't work after the next migration has completed.

**Comment by tomhughes** [comment]
> How will that code know if the first comment has been already deleted for a given note? It can't just be `all_comments.drop(1)` for example, since that won't work after the next migration has completed.

I was imagining deleting the first comment only if it was an `opened` comment and on top of that we could also require author and/or text to be the same.

**Comment by nenad-vujicic** [comment]
> I would also interested in what the next step will be.

At the moment, the plan is following:

1. Perform data migration
2. If we succeeded to migrate all notes, remove optional using records from first special comment, otherwise we'll have to leave some fallback mechanism
3. Create DB table `old_notes` which will store note versions and update Note model with appropriate associations
  - This will be decomposed into several PRs
  - Perhaps this is good place to discuss how to apply #5619 to notes tables (perhaps use `notes_versions` instead of `old_notes`?)
4. Add creating versions on notes actions (create, reopen, close, ..)
5. Populate `note_versions` from `note_comments`
6. Create helper routine which will generate list of comments (i.e. discussion, used in left side-bar, rss feed, xml / json files, ..) from `note_versions` table and from `note_comments` table (but using only note comments with non-blank bodies
7. Replace displaying / using discussions with helper routine
8. Remove column `event` from `note_comments` table and comments with blank bodies
9. .. continue with adding note tags ..

I'm aware this plan is full of flaws and there are multiple points which should be further discussed / defined, so everyone's comment is very welcome!

File: db/migrate/20250212160355_backfill_note_descriptions.rb (hunk 1 above)
Comment (nenad-vujicic) [reply db/migrate/20250212160355_backfill_note_descriptions.rb:24]: We have proposal for supporting notes with a missing opening comment (and generally all comments) in #5712. Do we have other special cases we should support before this?

**Comment by nenad-vujicic** [comment]
@tomhughes do you think we can continue with current version of this PR or we should try with more robust data-migration proposed [here](https://github.com/openstreetmap/openstreetmap-website/pull/5667#issuecomment-2659898440)?

**Comment by tomhughes** [comment]
Well we don't currently have the infrastructure in place for that method, and I'm not sure this change is large enough to really need it.

The big problem with that approach is that it doesn't force the migration to happen, so you have to document that the job needs to be run when upgrading past a certain point and then hope that everybody using the code base reads that document and runs the job at the right time.

**Comment by tomhughes** [comment]
I think this is fine in this case, so I'm going to go ahead and merge it. Thanks for all the work.
